system-image-upgrader: implement support for LVM migration

NotKit opened the pull request:
If enabled by setting ubuntu.use_lvm=1 property, it will attempt to slightly shrink userdata and convert it in-place to LVM logical volume.

Then the system partition is going to be migrated to LVM as well and expanded to accommodate for a larger rootfs on older devices.

Creating this as a draft for now so it can be reviewed and discussed.

NotKit committed:
system-image-upgrader: implement support for LVM migration

If enabled by setting ro.systemimage.use_lvm=true property, it will attempt
to slightly shrink userdata and convert it in-place to LVM logical
volume.

Then the system partition is going to be migrated to LVM
as well and expanded to accommodate for a larger rootfs on older devices.

Change-Id: I2006b328a50d6808254102ba1083b9d96efe1151

NotKit commented on ubports/system-image-upgrader:
```diff
@@ -256,6 +256,156 @@ factory_wipe() {
+    else
+        # Legacy system partition
+        if [ "$operation" == "format" ]; then
+            echo "Setting up system partition with LVM (fresh install)"
+            # Fresh install - just create LVM structure, no need to migrate data
+            # Create physical volume
+            pvcreate "$system_device"
+            # Create volume group (may already exist from userdata)
+            vgcreate ubuntu "$system_device" 2>/dev/null || vgextend ubuntu "$system_device"
+            # Create logical volume with required size
+            required_extents=$((required_size_mb / 1))
+            lvcreate -l "$required_extents" -n rootfs ubuntu
+        else
+            echo "Migrating system partition to LVM (delta upgrade)"
```
It's used by the other parts of the script, so it would fail anyway if not supported.

NotKit commented on ubports/system-image-upgrader:
```diff
@@ -256,6 +256,156 @@ factory_wipe() {
+    fi
+
+    # LVM not set up yet - check raw partition size
+    system_device=$(grep "^[^#]" /etc/fstab | grep -e "\(/mnt\)*/system\(_root\)*" | cut -f 1 -d\ )
```
This matches the parsing method used elsewhere in the script, which had been tested to work in our case.

peat-psuwit commented on ubports/lvm-super-migrator:
```diff
@@ -0,0 +1,242 @@
+# Exit immediately if a command exits with a non-zero status
+set -e
+
+STARTING_PHYSICAL_EXTENT_ADDRESS=1048576  # 1MB
```
Is this address in bytes or in extents? It seems to be in bytes, but the variable name makes it sounds like it's in extent.

NotKit replied:
Does STARTING_PHYSICAL_EXTENT_OFFSET sound better?

peat-psuwit replied:
Hmm... this is going to be bikeshedding, but what about `FIRST_PHYSICAL_EXTENT_OFFSET` or maybe `FIRST_PHYSICAL_EXTENT_OFFSET_BYTES`.

peat-psuwit commented on ubports/lvm-super-migrator:
```diff
@@ -0,0 +1,242 @@
+    # Append the segments
+    cat "$tmp_segments" >> /tmp/vgcfgrestore.txt
+
+    # Close the structure
+    cat >> /tmp/vgcfgrestore.txt << EOF
+  }
+}
+EOF
```
Command substitution also work under `<<EOF` segment. So this code can become:

```suggestion
  logical_volumes {
    $(cat "$tmp_segments")
  }
}
EOF
```

peat-psuwit commented on ubports/lvm-fs-migrator:
```diff
@@ -0,0 +1,246 @@
+    echo "Reserving $RESERVED_SIZE_MB MB for additional logical volumes"
+fi
+
+# Resize filesystem
+echo "Resizing filesystem..."
+resize_filesystem "$DEVICE_PATH" "$FILESYSTEM_SIZE"
+
+# Copy header to end
+HEADER_OFFSET="$FILESYSTEM_SIZE"
+echo "Copying 1MB header to offset: $HEADER_OFFSET"
+copy_header "$DEVICE_PATH" 0 "$HEADER_OFFSET"
```
Shouldn't FS size rounding to extent boundary happen here instead of in `resize_filesystem()`? As it is, if the resized filesystem is (somehow) not on 1MB boundary, then the header won't be too.

peat-psuwit commented on ubports/lvm-fs-migrator:
In general for this file: I think various geometry calculations should happen in single place and not scattered around various functions. This would reduce errors and make it easier to follow.

peat-psuwit commented on ubports/system-image-upgrader:
```diff
@@ -256,6 +256,206 @@ factory_wipe() {
+    fi
+}
+
+migrate_device_to_lvm() {
```
Given that migrating `super` partition is less risky than migrating `userdata` (the former involves rewriting only metadata, while the latter involves shuffling real data around), do you think it makes sense to leave `userdata` partition alone, if migrating `super` only gives suffice rootfs space?

NotKit replied:
It depends on the device. My intention was migrating `userdata` on OP5/6, where there is no `super` partition in the first place. I'm still not convinced we want to migrate `super` in the first place, I included the scripts to MR for the sake of completion, but it should not be enabled by default.

peat-psuwit replied:
I see. So on devices with `super` partition you plan to keep using Android's dynamic partition format?

NotKit replied:
It depends on a device. If the `super` partition is large enough to contain a vendor and an enlarged rootfs, the easiest option is to keep using that. However I think it's a matter of time till we encounter a device with `super` partition being too small and then it might be reasonable to migrate it, even if that means breaking fastbootd commands.

NotKit replied:
I pushed the changes to resize the logical partition in super to https://github.com/ubports/halium_bootable_recovery/tree/halium-13.0-lvm-super. Do you think it should be a separate MR? I would probably to cooperate with someone with FP4/FP5 to test it properly.

peat-psuwit replied:
Well, with the way it's integrated with the rest of this PR, I guess it can live under this PR.

peat-psuwit commented on ubports/system-image-upgrader:
Do you think it makes sense to rename the Android property away from `ubuntu` name and to maybe `system-image`?

How about LVM VG name? I guess `system-image` doesn't sound quite right, but, I don't know, `ubports` or `halium`?

peat-psuwit replied:
> Do you think it makes sense to rename the Android property away from ubuntu name and to maybe system-image?

Actually it probably should be `ro.system-image.[...]`.

NotKit replied:
Would `ro.system-image.` or `ro.system_image.` be preferred?

peat-psuwit replied:
Hmm... I don't know. I think I see both in `getprop` result on e.g. FP5 so both works, but I think I see the underscore variant more often. I'll leave this up to you.

NotKit replied:
I would probably go with `ro.systemimage` actually. Using `-` doesn't look nice for longer property names (from my view at least), and mixing dashes and underscores in a single name is a recipe for hardly noticeable typos to happen.

peat-psuwit commented on ubports/system-image-upgrader:
```diff
@@ -256,6 +256,206 @@ factory_wipe() {
+        # Use temporary VG name to avoid conflict with existing 'ubuntu' VG from userdata
+        temp_vg_name="ubuntu_super"
+
+        # Convert Android dynamic partitions to LVM structure
+        /system/bin/lvm-super-migrator --device="$super_partition" --vg-name="$temp_vg_name"
```
Note to self (?): this will probably affects looking up e.g. `vendor` partition in `lxc-android-config` as well.

fredldotme replied:
I am guessing it would also break fastbootd's ability to flash vendor images into `super`.

fredldotme commented:
How would we go about bind-mounting the fake `/cache` mountpoint? It's a prerequisite for the upgrader to be triggered by `recovery`, otherwise the recovery won't find the `ubuntu.command` file.

fredldotme commented:
Thinking of this again, we might need to make the recovery call into `lvm` commands before proceeding to the system-image-upgrader. The recovery has always been responsible for mounting `/data` and appropriate `/cache` bind-mounts (through init.rc hacks), so that it can figure out whether to run the upgrader or not. Unless re-architecting completely we must keep this intact.

NotKit commented:
> Thinking of this again, we might need to make the recovery call into `lvm` commands before proceeding to the system-image-upgrader. The recovery has always been responsible for mounting `/data` and appropriate `/cache` bind-mounts (through init.rc hacks), so that it can figure out whether to run the upgrader or not. Unless re-architecting completely we must keep this intact.

The `setup_fake_cache.sh` script is not even a part of this repo sadly, but gets copied from device to device. It's responsible for mounting `/data` and `/cache` initially and would need to be adjusted to call into `lvm` commands/account for the different paths.

fredldotme commented:
We should unify the `/data` and `/cache` handling then, maybe by bringing it into the main recovery code base, since this is a major change to the filesystem handling and easy to get wrong for beginner porters.

NotKit commented:
> We should unify the `/data` and `/cache` handling then, maybe by bringing it into the main recovery code base, since this is a major change to the filesystem handling and easy to get wrong for beginner porters.

Agree here. It shouldn't be even device-specific at this point. Correct me if wrong for `suzu`, but I don't think we have any devices left that use the actual cache partition for storing OTAs, it's always mounted from data.

fredldotme commented:
> > We should unify the `/data` and `/cache` handling then, maybe by bringing it into the main recovery code base, since this is a major change to the filesystem handling and easy to get wrong for beginner porters.
> 
> Agree here. It shouldn't be even device-specific at this point. Correct me if wrong for `suzu`, but I don't think we have any devices left that use the actual cache partition for storing OTAs, it's always mounted from data.

`suzu` was the first device to ship such a bind-mount hack for `/cache`, since then all devices needed a similar solution.

NotKit committed:
partitions_tools_extra: implement lpresize tool

Use the same library calls as fastbootd to provide CLI interface
to resize logical partitions inside Android super partition.

Change-Id: Ibe1cdd9fda75a0db76fa22a2d41f2b7b4268aad2

NotKit committed:
system-image-upgrader: attempt to resize system partition in super

When ro.systemimage.system_partition_size is set and LVM mode is not
enabled, attempt to expand the system partition inside the Android
dynamic super partition.

It's going to be up to the porter to check there is enough free
space in super to accomodate the required system partition size,
as otherwise the process would fail and prevent the update.

Change-Id: Ib473ba8c2680dbd39091dda34aa9f20ed6ba5e72

NotKit committed:
init: run vgscan to create LVM device nodes on boot

Change-Id: I50bcb45f76ce288587ce30cde56d80e3c10e6575

NotKit commented:
@peat-psuwit @fredldotme which LVM volume group name would we prefer? `ubuntu` or `ubports`? I want to add a [patch](https://github.com/Halium/hybris-patches/commit/e5294e973b8a67905dfde12b2f58a3e957497cfb) in Halium to override userdata block device in fstab if LVM setup is present, so it would be good to settle on the name.

peat-psuwit commented:
I'm inclined to say `systemimage`, but between `ubuntu` and `ubports` I would choose `ubports`.

NotKit commented:
Personally I'd find `ubports--rootfs` and `ubports--userdata` more meaningful than `systemimage` at least.

peat-psuwit commented on Android.bp:
```diff
@@ -196,6 +196,7 @@ cc_binary {
         "e2fsck.recovery",
         "gpg",
         "install-system",
+        "lvm",
```
I assume LVM is added to the tree elsewhere?

NotKit replied:
Should be added to @Halium/android `snippets/ubports.xml`, yes.

peat-psuwit replied:
Done in https://github.com/Halium/android/pull/100

peat-psuwit commented on partition_tools_extra/lpresize.cc:
```diff
@@ -0,0 +1,201 @@
+            LOG(ERROR) << "Failed to write metadata for slot " << i;
+            return false;
+        }
+    }
```
So you read the metadata from slot `slot_number`, and then write the updated metadata to all metadata slots?

NotKit replied:
Yes, I don't think we even handle slots properly in parse-android-dynparts, so it was just "to be safe". But now thinking of it again, if there is a mismatch between the slot number recovery and halium-boot ramdisk use, we're already in trouble, so I could remove this.

peat-psuwit replied:
But what usually defines which metadata slot to be used? Is this something used in virtual A/B scheme or something?

NotKit replied:
https://source.android.com/docs/core/ota/dynamic_partitions/ab_launch#background

According to Google's documentation, 0 is for _a and 1 is for _b essentially (but to make it complicated, they can still contain the partitions from the previous update, and the new virtual A/B scheme supports mapping a part of userdata to store the blocks of the old partition during OTA update in COW mode).

NotKit committed:
system-image-upgrader: enlarge rootfs filesystem on mount if needed

Handle the case of the filesystem size being smaller than
the underlying block device on delta upgrades, which is rare, but
could happen e. g. with some Volla factory images or after an
incomplete resize operation previously.

Change-Id: Id82291dcb54747dbad0cc45826012a1768de70de

peat-psuwit commented on ubports/system-image-upgrader:
```diff
@@ -256,6 +274,274 @@ factory_wipe() {
+    partition_name="system"
+    if [ -n "$slot_suffix" ]; then
+        partition_name="system${slot_suffix}"
+    fi
```
Nit:

```suggestion
    slot_suffix=$(getprop ro.boot.slot_suffix) # May be empty.
    partition_name="system${slot_suffix}"
```

NotKit replied:
Thanks

peat-psuwit commented:
Merging under the understanding that the LVM side has not been fully tested. The Android super partition side has been tested on FP5, and does work. So it's decided to have this merged to unblock Android super partition.
